### PR TITLE
use `torch.testing.assert_equal` in `TestCase.assertEqual`

### PR DESCRIPTION
**Pull request by pmeier**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #69751
* __->__ #67796

Supersedes #58981.

cc @mruberry

Differential Revision: [D33542994](https://our.internmc.facebook.com/intern/diff/D33542994)

### CODE REVIEW

File: test/test_fx.py [hunk 1]
```diff
@@ -2529,15 +2529,15 @@ def test_profiler_ranges_side_effect(self):
             if node.op == 'call_function':
                 found_targets.setdefault(node.target)
         self.assertEqual(
-            found_targets.keys(), [torch.ops.profiler._record_function_enter, torch.ops.profiler._record_function_exit])
+            found_targets.keys(), {torch.ops.profiler._record_function_enter, torch.ops.profiler._record_function_exit})
```
Comment (pmeier) [inline test/test_fx.py:2532]: Not sure if I'm missing anything here, but comparing the keys of a dictionary, which are unordered by definition, to an ordered sequence, is a recipe for a flaky test. @jamesr66a do we actually want the test the order here?

File: test/test_nn.py [hunk 2]
```diff
@@ -4022,7 +4022,7 @@ def test_spectral_norm(self):
         self.assertFalse('weight' in m._buffers)
         self.assertFalse('weight' in m._parameters)
         # it should also be sharing storage as `weight_orig`
-        self.assertEqual(m.weight_orig.storage(), m.weight.storage())
+        self.assertEqual(m.weight_orig.storage()._cdata, m.weight.storage()._cdata)
```
Comment (pmeier) [inline test/test_nn.py:4025]: From the comment I get that we actually want to make sure that both tensors share the same memory. The same C++ storage can have multiple Python identities so we need to make sure that they point to the same memory location.

File: test/distributed/test_data_parallel.py [hunk 3]
```diff
@@ -382,7 +382,7 @@ def test_data_parallel_sparse(self):
                 self.assertEqual(out.get_device(), dev_id[0])
                 self.assertEqual(out, expected_out)
                 for expected, param in zip(expected_grads, l.parameters()):
-                    self.assertEqual(param.grad, expected)
+                    self.assertEqual(param.grad.coalesce(), expected.coalesce())
```
Comment (pmeier) [inline test/distributed/test_data_parallel.py:385]: The old implementation of `assertEqual` coalesced by default. We opted to not do that in `assert_close` because it can be the reason for subtle bugs. So far there are only 2 tests that need fixing, which should be reasonable. If we do want the relaxed behavior for `assertEqual`, we can also fairly easily achieve that by overriding the attribute equalization on the tensor pair.

File: test/test_fx.py (hunk 1 above)
Comment (jamesr66a) [reply test/test_fx.py:2532]: @pmeier PyTorch requires at minimum Python 3.6. Dictionaries are insertion-ordered by implementation in CPython 3.6 (https://stackoverflow.com/q/39980323) and this is codified as language standard in 3.7+ (https://mail.python.org/pipermail/python-dev/2017-December/151283.html). Thus, there is no risk of flakiness here.

Comment (pmeier) [reply test/test_fx.py:2532]: Thanks for the info, I didn't know that. Still, it makes a difference for the testing: the old implementation of `assertEqual` checked for iterables, but treated them as sequences. The new implementation will no longer do that. Meaning, both inputs either have to be set-like if the order doesn't matter or sequences if it does:

```python
>>> d = dict(foo="bar", baz=0)
>>> d.keys() == ["foo", "baz"]
False
>>> d.keys() == {"foo", "baz"}
True
>>> list(d.keys()) == ["foo", "baz"]
True
```

**Comment by pmeier** [comment]
@JackCaoG It seems that `isclose` (or better the ops used in) do not work on [XLA](https://circleci.com/gh/pytorch/pytorch/16694912) in some cases. Could you have a look?

**Comment by JackCaoG** [comment]
@miladm Can you take a look at this one? Thanks!

File: test/distributed/test_data_parallel.py (hunk 3 above)
Comment (mruberry) [reply test/distributed/test_data_parallel.py:385]: cc @cpuhrsch @IvanYashchuk @pearu

File: test/quantization/core/test_quantized_tensor.py [hunk 4]
```diff
@@ -813,8 +813,13 @@ def test_clone(self):
                 [numel], scale=scale, zero_point=zero_point,
                 device=device, dtype=dtype)
             per_channel_quantized = torch._empty_per_channel_affine_quantized(
-                [numel], scales=torch.tensor([scale]), zero_points=torch.tensor([zero_point]), axis=0,
-                device=device, dtype=dtype)
+                [numel],
```
Comment (mruberry) [inline test/quantization/core/test_quantized_tensor.py:816]: What's the change here?

File: test/test_autograd.py [hunk 5]
```diff
@@ -5800,10 +5800,17 @@ def test(get_input, cuda, pin_memory):
                 self.assertEqual(a, y.grad_fn._saved_other)
                 self.assertEqual(a.dtype, y.grad_fn._saved_self.dtype)
                 self.assertEqual(a.layout, y.grad_fn._saved_self.layout)
-                if y.is_sparse:
+                if a.is_sparse:
```
Comment (mruberry) [inline test/test_autograd.py:5803]: cc @IvanYashchuk @cpuhrsch @pearu 

Why the change from `y` to `a` in the conditional?

File: test/test_cuda.py [hunk 6]
```diff
@@ -79,7 +79,7 @@ def make_sparse_tensor(t, n, *sizes):
         torch.cat([torch.LongTensor(1, n).random_(s) for s in sizes], 0))
     v = tensor._values()
     v = v.new(n).copy_(torch.randn(n))
-    return t(i, v, torch.Size(sizes))
+    return t(i, v, torch.Size(sizes)).coalesce()
```
Comment (mruberry) [inline test/test_cuda.py:82]: Note to our sparse colleagues (@cpuhrsch, @pearu, @IvanYashchuk): sparse tests in test_cuda.py should probably be moved into test_sparse.py

File: torch/testing/_comparison.py [hunk 7]
```diff
@@ -275,11 +275,20 @@ def __init__(
         self._unknown_parameters = unknown_parameters
 
     @staticmethod
-    def _check_inputs_isinstance(*inputs: Any, cls: Union[Type, Tuple[Type, ...]]):
-        """Checks if all inputs are instances of a given class and raise :class:`UnsupportedInputs` otherwise."""
+    def _check_inputs_isinstance(
+        *inputs: Any, cls: Union[Type, Tuple[Type, ...]], not_cls: Optional[Union[Type, Tuple[Type, ...]]] = None
+    ) -> None:
+        """Checks if all inputs are instances of a given class and and optionally not of another.
```
Comment (mruberry) [inline torch/testing/_comparison.py:281]: An interesting extension (that might be more logically separated into its own function)

File: torch/testing/_comparison.py [hunk 8]
```diff
@@ -803,6 +824,8 @@ def originate_pairs(
     expected: Any,
     *,
     pair_types: Sequence[Type[Pair]],
+    sequence_types: Tuple[Type, ...] = (collections.abc.Sequence,),
+    mapping_types: Tuple[Type, ...] = (collections.abc.Sequence,),
```
Comment (mruberry) [inline torch/testing/_comparison.py:828]: Not `abc.Mapping` for `mapping_types`?

File: torch/testing/_comparison.py [hunk 9]
```diff
@@ -866,9 +901,24 @@ def originate_pairs(
                 id=id,
             )
 
+        keys: Collection = actual_keys
+        # Since the origination aborts after the first failure, we try to be deterministic
```
Comment (mruberry) [inline torch/testing/_comparison.py:905]: Neat

File: torch/testing/_comparison.py [hunk 10]
```diff
@@ -847,10 +872,20 @@ def originate_pairs(
...
+                    sequence_types=sequence_types,
+                    mapping_types=mapping_types,
+                    id=(*id, idx),
+                    **options,
+                )
+            )
         return pairs
 
-    elif isinstance(actual, collections.abc.Mapping) and isinstance(expected, collections.abc.Mapping):
+    elif isinstance(actual, mapping_types) and isinstance(expected, mapping_types):
```
Comment (mruberry) [inline torch/testing/_comparison.py:888]: Won't this never be hit by default because sequence types and mapping types are the same?

File: torch/testing/_comparison.py [hunk 11]
```diff
@@ -884,7 +934,13 @@ def originate_pairs(
 
 
 def assert_equal(
-    actual: Any, expected: Any, *, pair_types: Sequence[Type[Pair]] = (ObjectPair,), **options: Any
+    actual: Any,
+    expected: Any,
+    *,
+    pair_types: Sequence[Type[Pair]] = (ObjectPair,),
+    sequence_types: Tuple[Type, ...] = (collections.abc.Sequence,),
+    mapping_types: Tuple[Type, ...] = (collections.abc.Mapping,),
```
Comment (mruberry) [inline torch/testing/_comparison.py:942]: See line 828

File: torch/testing/_internal/common_utils.py [hunk 12]
```diff
@@ -787,6 +795,25 @@ def _check_module_exists(name: str) -> bool:
     np.complex128 : torch.complex128
 }
 
+
+# numpy dtypes like np.float64 are not instances, but rather classes. This leads to rather absurd cases like
```
Comment (mruberry) [inline torch/testing/_internal/common_utils.py:799]: Nice comment

File: torch/testing/_internal/common_utils.py [hunk 13]
```diff
@@ -1309,30 +1336,148 @@ def check_if_enable(test: unittest.TestCase):
...
-    # non-CUDA (CPU, for example) float16 -> float32
-    # TODO: update this when isclose is implemented for CPU float16
-    if (compare_dtype is torch.float16 and
-        (a.device != b.device or a.device.type != 'cuda' or
-            b.device.type != 'cuda')):
-        compare_dtype = torch.float32
-
-    return compare_dtype
+
+class RelaxedBooleanPair(BooleanPair):
```
Comment (mruberry) [inline torch/testing/_internal/common_utils.py:1340]: Would you elaborate on this?

How bad would not supporting this be?

File: torch/testing/_internal/common_utils.py [hunk 14]
```diff
@@ -1309,30 +1336,148 @@ def check_if_enable(test: unittest.TestCase):
...
+                    f"Got {numel} elements instead.",
+                    id=id,
+                )
+
+            return bool(bool_like.item())
+        else:
+            return super()._to_bool(bool_like, id=id)
+
+
+class RelaxedNumberPair(NumberPair):
```
Comment (mruberry) [inline torch/testing/_internal/common_utils.py:1374]: Same question -- could we get away without supporting these relaxes variants?

File: torch/testing/_internal/common_utils.py [hunk 15]
```diff
@@ -1309,30 +1336,148 @@ def check_if_enable(test: unittest.TestCase):
...
+        # but will be picked up here since issubclass(bool, int)
+        elif isinstance(number_like, bool):
+            return int(number_like)
+        elif isinstance(number_like, Enum):
+            return int(number_like)  # type: ignore[call-overload]
+        else:
+            return super()._to_number(number_like, id=id)
+
+
+class TensorOrArrayPair(TensorLikePair):
```
Comment (mruberry) [inline torch/testing/_internal/common_utils.py:1430]: Comment needed here -- this is to support ndarray vs tensor comparisons, right?

File: torch/testing/_internal/common_utils.py [hunk 16]
```diff
@@ -1309,30 +1336,148 @@ def check_if_enable(test: unittest.TestCase):
...
+    def _process_inputs(self, actual, expected, *, id, allow_subclasses):
+        self._check_inputs_isinstance(actual, expected, cls=(torch.Tensor, np.ndarray))
+
+        actual, expected = [self._to_tensor(input, id=id) for input in (actual, expected)]
+        for tensor in (actual, expected):
+            self._check_supported(tensor, id=id)
+        return actual, expected
+
+
+class UnittestPair(Pair):
```
Comment (mruberry) [inline torch/testing/_internal/common_utils.py:1445]: Comments for all these classes

File: torch/testing/_internal/common_utils.py [hunk 17]
```diff
@@ -1309,30 +1336,148 @@ def check_if_enable(test: unittest.TestCase):
...
+        type_name = self.TYPE_NAME or (self.CLS if isinstance(self.CLS, type) else self.CLS[0]).__name__
+        return self._make_error_meta(AssertionError, f"{type_name.title()} comparison failed: {msg}")
+
+
+class StringPair(UnittestPair):
+    CLS = string_classes
+    TYPE_NAME = "string"
+
+
+class SetPair(UnittestPair):
```
Comment (mruberry) [inline torch/testing/_internal/common_utils.py:1470]: Do we not handle sets in assert_close?

File: torch/testing/_internal/common_utils.py (hunk 13 above)
Comment (mruberry) [reply torch/testing/_internal/common_utils.py:1340]: Also this should probably add a note about the relationship between these pairs, assertEqual, and assertClose so future maintainers (including ourselves!) understand why these pairs are here and how the best way to modify in the future

File: torch/testing/_internal/common_utils.py [hunk 18]
```diff
@@ -1767,304 +1912,77 @@ def compare_with_numpy(self, torch_fn, np_fn, tensor_like,
...
+            self,
+            x,
+            y,
+            msg: Optional[str] = None,
+            *,
+            atol: Optional[float] = None,
+            rtol: Optional[float] = None,
+            equal_nan=True,
+            exact_dtype=True,
+            exact_device=False,
```
Comment (mruberry) [inline torch/testing/_internal/common_utils.py:1930]: Let's keep the TODO comment:

```
exact_device=False,  # TODO: default this to True
```

File: torch/testing/_internal/common_utils.py [hunk 19]
```diff
@@ -1767,304 +1912,77 @@ def compare_with_numpy(self, torch_fn, np_fn, tensor_like,
...
+            exact_layout=False,
+            exact_stride=False,
+            exact_is_coalesced=False
+    ):
+        # Hide this function from `pytest`'s traceback
+        __tracebackhide__ = True
```
Comment (mruberry) [inline torch/testing/_internal/common_utils.py:1936]: Nice update

File: torch/testing/_internal/common_utils.py [hunk 20]
```diff
@@ -1767,304 +1912,77 @@ def compare_with_numpy(self, torch_fn, np_fn, tensor_like,
...
+
+        if any(
```
Comment (mruberry) [inline torch/testing/_internal/common_utils.py:1938]: This special case needs a comment

File: torch/testing/_internal/common_utils.py [hunk 21]
```diff
@@ -1767,304 +1912,77 @@ def compare_with_numpy(self, torch_fn, np_fn, tensor_like,
...
+            isinstance(input, np.ndarray) and not has_corresponding_torch_dtype(input.dtype) for input in (x, y)
+        ):
+            def to_list(input):
+                return input.tolist() if isinstance(input, (torch.Tensor, np.ndarray)) else list(input)
+
+            x = to_list(x)
+            y = to_list(y)
+        elif isinstance(x, torch.Tensor) and isinstance(y, Sequence):
```
Comment (mruberry) [inline torch/testing/_internal/common_utils.py:1946]: Comments here too

Readability nit: maybe keep `x` on the left and `y` on the right? 

Follow-up idea: `actual` and `expected` might be more readable names than `x` and `y` but nbd

**Review by mruberry — COMMENTED** [review]
Overall looks really good, @pmeier. I made some comments inline for your review and the failing tests need to be resolved; also pinged in our sparse experts to take a look at the sparse test updates

File: test/quantization/core/test_quantized_tensor.py (hunk 4 above)
Comment (pmeier) [reply test/quantization/core/test_quantized_tensor.py:816]: The number of elements of `scales` and `zero_points` did not match the values. Before we only compared the values so this worked out. Now we compare the `.dequantize`'d version.

File: test/test_autograd.py (hunk 5 above)
Comment (pmeier) [reply test/test_autograd.py:5803]: Consistency. Below we can no longer check for `y` so I thought always checking for `a` might be more clear. No strong opinion, can revert.

File: test/test_fx.py [hunk 22]
```diff
@@ -2010,11 +2010,11 @@ def test_reassign_args_kwargs_uses(self):
 
         # zed = z + z + z -> zed = z + z + x
         zed.node.args = (zed.node.args[0], x.node)
-        self.assertEqual(x.node.users.keys(), [z.node, zed.node])
+        self.assertEqual(list(x.node.users.keys()), [z.node, zed.node])
```
[the remark this answers is not in the document]
Comment (pmeier) [reply test/test_fx.py:2013]: See https://github.com/pytorch/pytorch/pull/67796#discussion_r745665381.

File: torch/testing/_comparison.py (hunk 8 above)
Comment (pmeier) [reply torch/testing/_comparison.py:828]: Good catch, this is wrong here.

File: torch/testing/_internal/common_utils.py (hunk 17 above)
Comment (pmeier) [reply torch/testing/_internal/common_utils.py:1470]: The old implementation had support for it:

https://github.com/pytorch/pytorch/blob/e3bcf64ff84f8e96839e39056b3b90d1bd1f8bbe/torch/testing/_internal/common_utils.py#L2025-L2028

 I'll remove it temporary and see how many tests break.

File: torch/testing/_internal/common_utils.py (hunk 21 above)
Comment (pmeier) [reply torch/testing/_internal/common_utils.py:1946]: Agreed. As long as no one is doing `self.assertEqual(x=..., y=...)` we can rename. I'll try in a follow-up PR.

File: test/test_autograd.py [hunk 23]
```diff
@@ -5800,10 +5800,17 @@ def test(get_input, cuda, pin_memory):
...
                     y = y.to_dense()
                 y.sum().backward()
-                self.assertEqual(2 * a, a.grad)
+
+                actual = 2 * a
+                expected = a.grad
+                if a.is_sparse:
+                    actual = actual.coalesce()
```
[the remark this answers is not in the document]
Comment (pmeier) [reply test/test_autograd.py:5810]: Yes. All changes made to tests in this PR are required to make the tests pass with the new infrastructure. Most of it is handled by the relaxed pairs, but some things, like comparing sparse tensors with different NNZ are no longer allowed. Let's keep the discussion in https://github.com/pytorch/pytorch/pull/67796#discussion_r745671190.

File: torch/testing/_internal/common_utils.py (hunk 13 above)
Comment (pmeier) [reply torch/testing/_internal/common_utils.py:1340]: > Would you elaborate on this?

The `BooleanPair` added in #67794 only handles `bool`'s and `np.bool_`'s. In our test suite there are many tests that compare `bool`'s against 0d tensors with dtype `torch.bool`. In some cases even something like `False` vs `torch.tensor(0, dtype=torch.int64)`.

> How bad would not supporting this be?

We would need to manually fix every single test that does something like I explained above. Worse, if we try to get rid of all `Relaxed*Pair`'s this PR will grow unreasonable large since we will need to touch a good portion of our test suite. Have a look at  #58981 where I had some automation in `assertEqual`, have already "fixed" a lot of tests, and still are not close to being done. 

Having the ability to use different pairs for our internal test suite was the major reason to refactor `assert_close` in the first place. If we want more strictness in the future, I suggest we add the relaxed pairs for now and phase them out later one by one.

File: test/distributed/test_data_parallel.py (hunk 3 above)
Comment (IvanYashchuk) [reply test/distributed/test_data_parallel.py:385]: I agree that an explicit call to `.coalesce()` is better for tests.
What are the possible subtle bugs when hiding `.coalesce()` inside the `assertEqual`? To me coalesced/uncoalesced is just for internal implementation convenience and it shouldn't matter if the sparse COO tensor has duplicates or not.

Comment (pmeier) [reply test/distributed/test_data_parallel.py:385]: > I agree that an explicit call to `.coalesce()` is better for tests.

Do you mean being explicit in the test or `assertEqual` should handle this?

> What are the possible subtle bugs when hiding `.coalesce()` inside the `assertEqual`? To me coalesced/uncoalesced is just for internal implementation convenience and it shouldn't matter if the sparse COO tensor has duplicates or not.

This behavior was added in #58844:

https://github.com/pytorch/pytorch/blob/33e9a0b5f6674bd429bda689534e1c987b38cf6e/torch/testing/_asserts.py#L293-L295

IIRC, @pearu wanted to have it. Maybe he can weigh in.

File: torch/testing/_internal/common_utils.py (hunk 14 above)
Comment (pmeier) [reply torch/testing/_internal/common_utils.py:1374]: See https://github.com/pytorch/pytorch/pull/67796#discussion_r750019366.

File: torch/testing/_internal/common_utils.py (hunk 17 above)
Comment (pmeier) [reply torch/testing/_internal/common_utils.py:1470]: Oh sorry, I think I misunderstood: no I don't think we should support sets in `assert_close`, because that would mean we need to somehow match tensors in two sets before we can do an actual numeric comparison. That is easy for sequences and mappings by index / key, but I don't think that is possible for unordered containers.

File: test/quantization/core/test_quantized_tensor.py [hunk 24]
```diff
@@ -165,7 +165,7 @@ def _test_qtensor_dynamic(self, device):
             scale, zero_pt = _calculate_dynamic_qparams(mat2quant, dtype, reduce_flag)
             q_s = torch.quantize_per_tensor(mat2quant, scale, zero_pt, dtype)
 
-            self.assertEqual(q_d, q_s)
+            self.assertEqual(q_d, q_s, rtol=1.3e-6, atol=1e-5)
```
Comment (pmeier) [inline test/quantization/core/test_quantized_tensor.py:168]: @dzdang could you have a look at this, since the test was only added a few days ago in #68004. For context, the current implementation checks the quantization attributes as well as the integer representation individually 

https://github.com/pytorch/pytorch/blob/33e9a0b5f6674bd429bda689534e1c987b38cf6e/torch/testing/_internal/common_utils.py#L1967-L1971

In #58926 we decided to only keep the check for the `.qscheme()`

https://github.com/pytorch/pytorch/blob/33e9a0b5f6674bd429bda689534e1c987b38cf6e/torch/testing/_asserts.py#L248-L253

and test the `.dequantize()`'d versions:

https://github.com/pytorch/pytorch/blob/33e9a0b5f6674bd429bda689534e1c987b38cf6e/torch/testing/_asserts.py#L175

I've used the `float32` default tolerances here, because that is the `dtype` you get after `.dequantize()`ing the tensors.

**Comment by pmeier** [comment]
Ping @miladm for looking into `torch.isclose` for XLA, which is the last test failure.

File: test/quantization/core/test_quantized_tensor.py (hunk 24 above)
Comment (dzdang) [reply test/quantization/core/test_quantized_tensor.py:168]: I'll look into this. This was a bootcamp task for me, and I'm not familiar with all that's in the codebase yet.

Comment (mruberry) [reply test/quantization/core/test_quantized_tensor.py:168]: @vkuzo might want to take a look, too, to get an idea for what this update will change in quantization testing

Comment (pmeier) [reply test/quantization/core/test_quantized_tensor.py:168]: After some more thought, if we keep the check for the `.dequantize()`'d versions, we probably need to relax the default tolerances. Currently they are treated as integers and thus `rtol = atol = 0`, but the values are compared as floating points.

Comment (pmeier) [reply test/quantization/core/test_quantized_tensor.py:168]: See #68548.

File: test/distributed/test_data_parallel.py (hunk 3 above)
Comment (pearu) [reply test/distributed/test_data_parallel.py:385]: The old assertEqual implementation was based on the following considerations:
1. there are different levels of equality tests: mathematical equality (e.g. comparing tensors with different layouts), representation equality (e.g. comparing tensors with the same layout), and internal representation equality (e.g. comparing uncoalesced state of COO tensors as defined by `_values()` and `_indices()`).
2. `assertEqual` mostly deals with representation equality and for all other layouts other than COO (so far), representation and internal representation equalities are identical.
3. Internal representation equality can be valuable for ensuring that internal optimizations function as expected. For instance, when constructing a COO tensor from in-place additions, one would want to ensure that the COO tensor stays uncoalesced in-between such operations. A failure to do so would mean a considerable performance hit. Notice that representation equality would not be able to detect this kind of regressions.

IIUC, it has been suggested here to implement only the representation equality for COO tensors, that is, `assertEqual` would always call `.coalesce()` before comparing the content of tensors. That would be fine with me (in the sense of still being able to test internal representation equality) iff `assertEqual` would support also tuple object containing tensors as items (unless it does already support that). If so, for internal representation equality tests one could introduce a utility function:
```python
def serialize(coo):
    return coo._indices(), coo._values()
```
and perform the internal representation equality test on COO tensors as follows:
```python
self.assertEqual(serialize(actual), serialize(expected))
```

Does the above make sense?

**Review by pearu — COMMENTED** [review]
Added the background of equality tests for COO tensors and a feature request for assertEqual to support tuples as inputs.

**Comment by wonjoolee95** [comment]
I'll be looking into the PyTorch/XLA build failure issue. I'll keep updates on this PR with findings.

File: torch/testing/_internal/common_methods_invocations.py [hunk 25]
```diff
@@ -10552,7 +10552,7 @@ def ref_pairwise_distance(input1, input2):
                         # NumPy's heaviside promotes bool to float16
                         DecorateInfo(unittest.expectedFailure,
                                      'TestBinaryUfuncs',
-                                     'test_reference_numerics_heavisidel',
+                                     'test_reference_numerics',
```
Comment (pmeier) [inline torch/testing/_internal/common_methods_invocations.py:10555]: This seems like a typo from #69439 cc @mruberry. Without this, the test is not properly skipped.

File: torch/testing/_internal/common_methods_invocations.py [hunk 26]
```diff
@@ -10549,11 +10549,6 @@ def ref_pairwise_distance(input1, input2):
                     supports_autograd=False,
                     # FIXME: heaviside does not accept scalar inputs
                     skips=(
-                        # NumPy's heaviside promotes bool to float16
-                        DecorateInfo(unittest.expectedFailure,
-                                     'TestBinaryUfuncs',
-                                     'test_reference_numerics_heavisidel',
-                                     dtypes=(torch.bool,)),
```
Comment (pmeier) [inline torch/testing/_internal/common_methods_invocations.py:10556]: The test name is malformed. It should be `test_reference_numerics` rather than `test_reference_numeric_heavisidel`. As a consequence, this test was never skipped and passed "unexpectedly" with the current behavior of `assertEqual`. This showed up, because at the time the comparison logic did not handle `torch.tensor(True)` vs `np.float16(1.0)` well. This is now fixed.

**Review by mruberry — APPROVED** [review]

**Comment by mruberry** [comment]
@mruberry has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33542994).

**Comment by mruberry** [comment]
The earlier part of this stack is landing now but there are some internal issues with this one we'll need to review. They are:

test_isclose_complex_cpu_complex{64, 128} are both failing with:

```
AssertionError: Booleans mismatch: False is not True
  File "test/testing/torch/testing/_internal/common_device_type.py", line 376, in instantiated_test
    result = test(self, **param_kwargs)
  File "test/testing/test_testing.py", line 203, in test_isclose_complex
    self._isclose_helper(tests, device, dtype, False, atol=.5, rtol=.5)
  File "test/testing/test_testing.py", line 57, in _isclose_helper
    self.assertEqual(actual.item(), expected)
  File "torch/testing/_internal/common_utils.py", line 2093, in assertEqual
    assert_equal(
  File "torch/testing/_comparison.py", line 1081, in assert_equal
    raise error_metas[0].to_error()
```

which is an odd trace without a great message so I'm not sure what's happening.

There are some additional internal failures whose error messages are so odd I'm not sure how the tests are passing today, like:

```
AssertionError: The values for attribute 'shape' do not match: torch.Size([0, 1, 2]) != torch.Size([0]).
```

These ultimately trace to the same line, too:

```
  File "test/testing/torch/testing/_comparison.py", line 1081, in assert_equal
    raise error_metas[0].to_error()
```

I'm creating an internal build to run these and hope to understand them better now.

File: test/test_testing.py [hunk 27]
```diff
@@ -311,12 +140,9 @@ def test_isclose_comparetensors_float(self, device, dtype):
 
         self._isclose_helper(tests, device, dtype, True)
 
-        self._comparetensors_helper(tests, device, dtype, True)
-
-    # Note: compareTensor will compare the real and imaginary parts of a
-    # complex tensors separately, unlike isclose.
+    @unittest.skipIf(IS_SANDCASTLE, "Skipping because doesn't work on sandcastle")
```
Comment (pmeier) [inline test/test_testing.py:143]: This test was already skipped by sandcastle through internal means. Due to the rename of the test the old skip is no longer in place and cannot be activated preemptively. Thus, we now publicly skip it.

**Review by pmeier — COMMENTED** [review]
After reviewing the failures, we found that while this PR caused them, they relied on internal behavior that should have failed in the first place and is only now surfaced. Examples:

1. `TestCase.assertEqual(torch.empty(0, 1, 2), [])`
2. `TestCase.assertEqual(unittest.mock.MagicMock(), {})`

They both relied on

https://github.com/pytorch/pytorch/blob/ff78c73286154ae704eca3bf447bb1f3e20ac830/torch/testing/_internal/common_utils.py#L1151-L1156

https://github.com/pytorch/pytorch/blob/ff78c73286154ae704eca3bf447bb1f3e20ac830/torch/testing/_internal/common_utils.py#L2139-L2145

to use an elementwise comparison of empty iterables. With this PR

1. comparing a list to a tensor will construct a tensor from the list, i.e. `torch.tensor([])`. This successively fails since the shapes don't match: `(0, 1, 2) != (0,)`.
2. being iterable is no longer valid ground to do elementwise comparisons. The inputs have to be ordered iterables, i.e. sequences. Since neither a dictionary nor a `unittest.mock.MagicMock` statisfy this, we fall back to a regular equality comparison with `==` which fails.

I added a comment below for one exception to the examples above.

**Comment by mruberry** [comment]
@mruberry has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33542994).

File: torch/testing/_comparison.py [hunk 28]
```diff
@@ -832,14 +839,14 @@ def _compare_regular_values_close(
     ) -> None:
         """Checks if the values of two tensors are close up to a desired tolerance."""
         actual, expected = self._promote_for_comparison(actual, expected)
-        mismatches = ~torch.isclose(actual, expected, rtol=rtol, atol=atol, equal_nan=equal_nan)
-        if not torch.any(mismatches):
+        matches = torch.isclose(actual, expected, rtol=rtol, atol=atol, equal_nan=equal_nan)
+        if torch.all(matches):
```
Comment (pmeier) [inline torch/testing/_comparison.py:843]: @mruberry This should resolve the performance degradation that you saw in the internal tests. The old `assertEqual` invoked `torch.allclose`

https://github.com/pytorch/pytorch/blob/26d54b4076fe6b6fd6e6c517643a19cfa5dcd7d9/torch/testing/_core.py#L96-L97

which is comparable to `torch.isclose` and `torch.all` here. 

The slowdown happened because we previously also inverted the array to include some mismatch statistics in the error message. Since we only need the mismatches in the error case, we can safely push the inversion there.

**Comment by mruberry** [comment]
@mruberry has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33542994).

**Comment by mruberry** [comment]
Unfortunately the internal tools aren't able to rebase this automatically. Rebasing should also move past all the failures that were in the base.

**Comment by mruberry** [comment]
@mruberry has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33542994).

**Comment by mruberry** [comment]
@mruberry has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33542994).
